### PR TITLE
GH-43688: [C++] Prevent Snappy from disabling RTTI when bundled

### PR DESCRIPTION
**Pull request by pitrou**
### Are there any user-facing changes?

* GitHub Issue: #43688

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1353,26 +1353,9 @@ macro(build_snappy)
...
-    # On macOS 10.13 we need to explicitly add <functional> to avoid a missing include error
-    # This can be removed once CRAN no longer checks on macOS 10.13
-    find_program(PATCH patch REQUIRED)
-    set(SNAPPY_PATCH_COMMAND ${PATCH} -p1 -i ${CMAKE_CURRENT_LIST_DIR}/snappy.diff)
-  else()
-    set(SNAPPY_PATCH_COMMAND)
-  endif()
+  # See comments in snappy.diff.
+  find_program(PATCH patch REQUIRED)
+  set(SNAPPY_PATCH_COMMAND ${PATCH} -p1 -i ${CMAKE_CURRENT_LIST_DIR}/snappy.diff)
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: @kou Do we require `patch` to be present on all platforms, especially Windows? Or should I make this more specific?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: Well, to answer myself: our emscripten builds don't have access to a `patch` executable...
https://github.com/ursacomputing/crossbow/actions/runs/10388720694/job/28764995095#step:7:826

**Comment by pitrou** [comment]
@github-actions crossbow submit -g r

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: And even some Linux builds such as `test-r-rocker-r-ver-latest` don't seem to have `patch` either. I'll try to make it optional.

**Comment by pitrou** [comment]
@github-actions crossbow submit *emscripten*

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Ok, unfortunately the patch is required. We either have to make sure that `patch` is available on all images, or do things differently :-/

**Comment by pitrou** [comment]
@github-actions crossbow submit *emscripten*

**Comment by pitrou** [comment]
@github-actions crossbow submit -g r

**Comment by pitrou** [comment]
@github-actions crossbow submit *emscripten*

**Comment by pitrou** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python -g cpp

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: Yes. `patch` is required as you noticed.
I think that another approach is better. See also: https://github.com/apache/arrow/issues/43688#issuecomment-2289907756

Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: OTOH, `patch` is useful and could be needed for other bundled builds.

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: I think that we should fix in upstream instead of using patch as much as possible :-)

Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1358]: I agree with this, but as you know Snappy rarely takes outside contributions...

**Comment by pitrou** [comment]
Closing as outdated, see the alternative and merged PR #43706

**Pull request closed without merging**